velox/common/caching/tests/SsdFileTest.cpp

```diff
@@ -427,17 +449,156 @@ TEST_F(SsdFileTest, checkpoint) {
+  initializeSsdFile(kSsdSize, checkpointIntervalBytes, true, true);
+  EXPECT_EQ(checkEntries(allEntries), 0);
+  EXPECT_EQ(ssdFile_->testingStats().readCheckpointErrors, 1);
+
+  // New entries can be written.
+  allEntries.clear();
+  for (auto startOffset = 0; startOffset <= kSsdSize - SsdFile::kRegionSize;
+       startOffset += SsdFile::kRegionSize) {
+    auto pins =
+        makePins(fileName_.id(), startOffset, 4096, 2048 * 1025, 62 * kMB);
+    ssdFile_->write(pins);
+    for (auto& pin : pins) {
+      EXPECT_EQ(ssdFile_.get(), pin.entry()->ssdFile());
+      allEntries.emplace_back(
+          pin.entry()->key(), pin.entry()->ssdOffset(), pin.entry()->size());
+    };
+    readAndCheckPins(pins);
+  }
+}
+
+TEST_F(SsdFileTest, recoverFromCheckpointWithChecksum) {
+  constexpr int64_t kSsdSize = 16 * SsdFile::kRegionSize;
+  const int32_t checkpointIntervalBytes = 5 * SsdFile::kRegionSize;
+  FLAGS_ssd_verify_write = true;
+
+  // Test if cache data can be recovered with different settings.
+  struct {
+    bool writeEnabled;
+    bool readVerificationEnabled;
+    bool dataRecoverable;
+
+    std::string debugString() const {
+      return fmt::format(
+          "writeEnabled {}, readVerificationEnabled {}, dataRecoverable {}",
+          writeEnabled,
+          readVerificationEnabled,
+          dataRecoverable);
+    }
+  } testSettings[] = {
+      {false, false, true},
+      {false, true, false},
+      {true, false, true},
+      {true, true, true}};
+
+  for (const auto& testData : testSettings) {
+    SCOPED_TRACE(testData.debugString());
+    // Initialize cache with checksum write enabled/disabled.
+    initializeCache(
+        kSsdSize, checkpointIntervalBytes, testData.writeEnabled, false);
+
+    // Populate the cache with some entries.
+    std::vector<TestEntry> allEntries;
+    for (auto startOffset = 0; startOffset <= kSsdSize - SsdFile::kRegionSize;
+         startOffset += SsdFile::kRegionSize) {
+      auto pins =
+          makePins(fileName_.id(), startOffset, 4096, 2048 * 1025, 62 * kMB);
+      ssdFile_->write(pins);
+      for (auto& pin : pins) {
+        EXPECT_EQ(ssdFile_.get(), pin.entry()->ssdFile());
+        allEntries.emplace_back(
+            pin.entry()->key(), pin.entry()->ssdOffset(), pin.entry()->size());
+      };
+    }
+    // All entries can be found.
+    EXPECT_EQ(checkEntries(allEntries), allEntries.size());
+
+    // Try reinitializing cache from checkpoint with read verification
+    // enabled/disabled.
+    ssdFile_->checkpoint(true);
+    initializeSsdFile(
+        kSsdSize,
+        checkpointIntervalBytes,
+        false,
+        testData.readVerificationEnabled);
+
+    // Check if cache data is recoverable as expected.
+    if (testData.dataRecoverable) {
+      EXPECT_EQ(checkEntries(allEntries), allEntries.size());
+    } else {
+      EXPECT_EQ(checkEntries(allEntries), 0);
+    }
+    cache_->shutdown();
+    memory::MemoryManager::testingSetInstance({});
```

Comment:
Why we need to reset memory manager here?

Reply:
If the allocation already have a cache registered, the new cache initialization will fail: https://github.com/facebookincubator/velox/blob/528193a79c3c6201e11042c5e6deccb628142a78/velox/common/memory/MallocAllocator.h#L34